### DISCUSSION TITLE
docs/guide/first-entity

### DISCUSSION THREAD

**Discussion by giscus** [mikro-orm/mikro-orm#5352] [Comments]
# docs/guide/first-entity

Setting up

https://mikro-orm.io/docs/guide/first-entity

**Comment by vitali-litvinovich** [comment]
This is amazing.

I've started software engineering 3 years ago, and back then I thought of these ORM as arbitrary, unclear. Only by working for 1.5 years with raw `mongodb` did I get to realise, how freaking time-saving, wonderful, beautiful, valuable this is.

Thanks for this guide, very straight-forward.

**Comment by omaerkhan** [comment]
Thanks for creating this - I've been writing SQL queries by hand for decades, tried a bunch of ORMs over the years and couldn't stick with any of them. This is the first time an ORM seems to just make sense to me...

I followed the steps above, but used a compound Primary Key (defined multiple Primary Keys on my entity) - and it seemed to work fine pretty much everywhere except:
- em.findOne followed by em.findOne both seem to run a SQL query (so not caching?)
- em.getReference doesn't seem to support compound PKs (I tried em.getReference(Entity, {key1: value, key2: value}) and it gives an error "No overload matches this call".

I do hope I'm doing something wrong because this seems really cool.

**Reply by B4nan** [reply]
> em.getReference doesn't seem to support compound PKs (I tried em.getReference(Entity, {key1: value, key2: value}) and it gives an error "No overload matches this call".

You need to pass a PK tuple into the method, not an object, so something like `em.getReference(Entity, [value1, value2])`, order must be the same as the order of PK properties in your entity definition.

> em.findOne followed by em.findOne both seem to run a SQL query (so not caching?)

This is the same as above likely. The query is skipped only if you search by the PK value directly, so `em.findOne(Entity, [value1, value2])`.

Note that for this to type check, you need to use the `PrimaryKeyProp` symbol in your entity definition.

https://mikro-orm.io/docs/composite-keys#primitive-types-only

**Reply by omaerkhan** [reply]
Thanks for the super fast n helpful reply :D

**Reply by omaerkhan** [reply]
@B4nan - FYI, I've got everything working... But the caching doesn't work even if I use a tuple - it still queries the DB for fresh data. I'm too new to know for sure that this is a bug, but it isn't bothering me much for now so I'm good to go. Thanks once again!

**Comment by amerkarim** [comment]
I'm having an issue, seems like it's related to how ts-node interacts with newer nodejs versions. 
While following this tutorial, I've added the code in server.ts according to the "Initializing the ORM" section, but I'm getting the following error:

 `triggerUncaughtException(
    ^
Error: Only URLs with a scheme in: file, data, and node are supported by the default ESM loader. On Windows, absolute paths must be valid file:// URLs. Received protocol 'd:'
    at throwIfUnsupportedURLScheme (node:internal/modules/esm/load:195:11)
    at defaultLoad (node:internal/modules/esm/load:78:3)
    at nextLoad (node:internal/modules/esm/hooks:785:28)
    at D:\Projects\ts-node-test\node_modules\ts-node\src\esm.ts:255:45
    at async addShortCircuitFlag (D:\Projects\ts-node-test\node_modules\ts-node\src\esm.ts:409:15)
    at async nextLoad (node:internal/modules/esm/hooks:785:22)
    at async AsyncLoaderHooksOnLoaderHookWorker.load (node:internal/modules/esm/hooks:410:20)
    at async handleMessage (node:internal/modules/esm/worker:251:18) {
  code: 'ERR_UNSUPPORTED_ESM_URL_SCHEME'
`

I'm on Windows 11, Node.js v24.13.0

**Reply by amerkarim** [reply]
After testing the same project on Ubuntu , seems like things are working correctly, looks like a Windows only issue.

**Reply by B4nan** [reply]
See #6670. This is resolved in v7.
